test/handler.js

```diff
@@ -169,6 +169,35 @@ describe('handler', () => {
                 done();
             });
         });
+
+        it('catches unhandled promise rejections when a promise is returned', (done) => {
+
+            const server = new Hapi.Server();
+            server.connection();
+
+            const asyncOperation = function () {
+
+                return new Promise((resolve, reject) => {
+
+                    throw new Error('This should be rejected...');
```

Comment:
I'd add a test for your  `setTimeout(() => throw new Error(), 1000);` too from above.